### PR TITLE
Shortfin llm no device syncs

### PR DESCRIPTION
**Pull request by stbaione**
Remove `await device0` in `Prefill/Decode` invocation. Device synchronizations are costly. Instead, we replace it with a `hack` that lets us check the host `device_array` and ensure that the data has been copied over.

For decode, instead of performing `req_count` transfers from device-to-host, we perform a single transfer and slice off of a single host array to ensure that the correct reqs have the correct logits.

### CODE REVIEW

**Review by monorimet — APPROVED** [review]
Thanks for adding the comment calling this out as a temporary workaround. LGTM.

File: shortfin/python/shortfin_apps/llm/components/batcher.py [hunk 1]
```diff
@@ -472,7 +472,7 @@ async def get_results(self, logits, req_count, device0):
             if req.return_host_array:
                 req.result_logits = logits_item.for_transfer()
                 req.result_logits.copy_from(logits_item)
-                await device0
+                check_host_array(req.result_logits)
             else:
                 req.result_logits = logits_item
             req.done.set_success()
```
Comment (zphoenixrises) [inline shortfin/python/shortfin_apps/llm/components/batcher.py:478]: You could probably change this to skip the loop for requests that have already completed

**Review by zphoenixrises — CHANGES_REQUESTED** [review]

File: shortfin/python/shortfin_apps/utils.py [hunk 2]
```diff
@@ -200,6 +201,44 @@ def convert_list_to_device_array(
...
+            )
+
+        dtype = dtype_map[host_dtype]
+        arr = np.frombuffer(array, dtype=dtype)
+        if not np.all(arr == 0):
+            check_1 = arr
+            check_2 = np.frombuffer(array, dtype=dtype)
+            if np.array_equal(check_1, check_2):
+                break
+    return
```
[the remark this answers is not in the document]
Comment (stbaione) [reply shortfin/python/shortfin_apps/utils.py:239]: I previously had an `await asyncio.sleep(0.05)` line in the loop, but it caused the benchmark latency to increase significantly. Can check again

File: shortfin/python/shortfin_apps/llm/components/batcher.py (hunk 1 above)
Comment (stbaione) [reply shortfin/python/shortfin_apps/llm/components/batcher.py:478]: Can you elaborate on what you mean by `already completed`? This is setting the results after the prefill invocation, then setting success for post-processing, so IIUC, it's already completed after we set the logits, which is what this loop does

File: shortfin/python/shortfin_apps/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AWoloszyn) [reply shortfin/python/shortfin_apps/utils.py:239]: If the gpu stalls for any reason while copying the data out (not sure what coherency flags are set on the buffer) you may exit the loop early without having all of the data.

**Comment by stbaione** [comment]
Closing PR. This was experimental and has proven to be too much of a hack
